### PR TITLE
py/objgenerator: Add __name__ to generator objects

### LINKED ISSUE #3293: Generator function has no __name__ attribute
Im in a situation when im required to save the name of the function from a decorator:

```
# ugly
def _hack_generator_name(fn):
    if not inspect.isgeneratorfunction(fn):
        return fn.__name__
    try:
        fn()
    except Exception as e:
        name = str(e)
        return name[:name.index("(")]


def optional_arg_decorator(fn):
    def wrapped_decorator(*args):
        if len(args) == 1 and callable(args[0]):
            return fn(args[0])
        else:
            def real_decorator(decoratee):
                return fn(decoratee, *args)
            return real_decorator

    return wrapped_decorator


@optional_arg_decorator
def httppost(fn, path=None, name=None):
    _decorator_list.append((_hack_generator_name(fn), path, Request.POST))
    return fn


@optional_arg_decorator
def httpget(fn, path=None, name=None):
    _decorator_list.append((_hack_generator_name(fn), path, Request.GET))
    return fn
```

But `fn.__name__ ` is missing from generator functions.

I solve this issue temporary by using an ugly hack (by calling the method and finding the name from the exception...)

### PR DESCRIPTION
**Pull request by same-id**
Fixes #3293

Hacktoberfest TLV

### CODE REVIEW

File: py/objgenerator.c [hunk 1]
```diff
@@ -230,6 +230,19 @@ STATIC const mp_rom_map_elem_t gen_instance_locals_dict_table[] = {
 
 STATIC MP_DEFINE_CONST_DICT(gen_instance_locals_dict, gen_instance_locals_dict_table);
 
+#if MICROPY_PY_GENERATOR_ATTRS
+STATIC void gen_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+	mp_obj_gen_instance_t *self = MP_OBJ_TO_PTR(self_in);
```
Comment (pfalcon) [inline py/objgenerator.c]: Incorrect indentation here and everywhere else, please see https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md

File: py/objgenerator.c [hunk 2]
```diff
@@ -238,4 +251,7 @@ const mp_obj_type_t mp_type_gen_instance = {
     .getiter = mp_identity_getiter,
     .iternext = gen_instance_iternext,
     .locals_dict = (mp_obj_dict_t*)&gen_instance_locals_dict,
```
Comment (pfalcon) [inline py/objgenerator.c]: A minor, but most types have .locals_dict as the last line, would be nice to follow that.

**Comment by dpgeorge** [comment]
It might be possible to implement this in a more compact way by reusing `fun_bc_attr()` (in objfun.c): point to this function in the generator type info, then modify that function to check the type of self_in.  If self_in is a generator type then extract the `self->code_state.fun_bc` value, otherwise just use self_in as that value.

If that works and doesn't increase the code size by much then I'd also suggest not having an additional config option and putting this feature under the existing MICROPY_PY_FUNCTION_ATTRS config option.

**Comment by jpochyla** [comment]
> If self_in is a generator type then extract the self->code_state.fun_bc value, otherwise just use self_in as that value.

`mp_obj_gen_instance_t` is defined in `objgenerator.c`, how would you access the code state from `objfun.c` in this case?

**Comment by dpgeorge** [comment]
It seems that this PR implements `__name__` for generator instances, rather than generating functions which what #3293 was reporting.  And support for `__name__` on generating functions was added in e2e22e3d7e29985579b2c91c639c71229422f349

**Pull request closed without merging**